tests/kernel/mem_protect/stackprot : input string is long stack overflow

The procedure followed to re-produce is -

Zephyr Commit ID:-
036dde2201ad60f2f1df2c73b62031a6967c629d
Config file is attached to this Report:
[kernel_mem_protect_stackprot_config.txt](https://github.com/zephyrproject-rtos/zephyr/files/1478228/kernel_mem_protect_stackprot_config.txt)


Build Procedure-

Cmake Build procedure -

$cd $ZEPHYR_BASE/tests/kernel/mem_protect/stackprot
 $mkdir build && cd build
 $cmake –DBOARD=quark_se_c1000_devboard..
 $make flash


Detailed description
====================
build passed: 'mkdir -p outdir-aetj-ijnd-quark_se_c1000_ss_devboard; rm -f outdir-aetj-ijnd-quark_se_c1000_ss_devboard/.config'

build passed: 'cmake -DPREBUILT_HOST_TOOLS= 
-DBOARD=quark_se_c1000_ss_devboard 
-DARCH=arc -DEXTRA_CPPFLAGS="-DTC_RUNID=:ijnd" 
-DEXTRA_CFLAGS="-Werror -Wno-error=deprecated-declarations" 
-DEXTRA_AFLAGS=-Wa,--fatal-warnings -DEXTRA_LDFLAGS=--fatal-warnings  -B"outdir-aetj-ijnd-quark_se_c1000_ss_devboard" -H"zephyr/tests/kernel/mem_protect/stackprot"'

build passed: 'mkdir -p outdir-aetj-ijnd-quark_se_c1000_devboard; rm -f outdir-aetj-ijnd-quark_se_c1000_devboard/.config'
build passed: 'cmake -DPREBUILT_HOST_TOOLS=zephyr-sdk-0.9.2/sysroots/x86_64-pokysdk-linux/usr/bin -DBOARD=quark_se_c1000_devboard -DARCH=x86
 -DEXTRA_CPPFLAGS="-DTC_RUNID=:ijnd"
 -DEXTRA_CFLAGS="-Werror -Wno-error=deprecated-declarations"
 -DEXTRA_AFLAGS=-Wa,--fatal-warnings
 -DEXTRA_LDFLAGS=--fatal-warnings  -
B"outdir-aetj-ijnd-quark_se_c1000_devboard" 
-H"zephyr/tests/booting/stub"'

build passed

deploy passed

Reset

console output: starting test - Test Stack Protection Canary
console output:
console output: Starts main
console output: Starts alternate_thread
console output: alternate_thread: Input string is too long and stack overflowed!
console output:
console output: alternate_thread: Stack ok
console output: alternate_thread: Stack ok
console output: alternate_thread: Stack ok
console output: alternate_thread: Stack ok
console output: alternate_thread: Stack ok
console output: alternate_thread: Stack ok
console output: Exception vector: 0x3, cause code: 0x6, parameter 0x0
console output: Address 0x206f6f74
console output: Current thread ID = 0xa8000450
console output: Faulting instruction address = 0x206f6f74
console output: Fatal fault in thread 0xa8000450! Aborting.
console output: FAIL - main.
console output:

```
Performing C SOURCE FILE Test check_fstack_protector_all failed with the following output:
Change Dir: /projects/zephyr/tests/kernel/mem_protect/stackprot/out/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/gmake" "cmTC_b2783/fast"
/usr/bin/gmake -f CMakeFiles/cmTC_b2783.dir/build.make CMakeFiles/cmTC_b2783.dir/build
gmake[1]: Entering directory '/projects/zephyr/tests/kernel/mem_protect/stackprot/out/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_b2783.dir/src.c.obj
/opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/arc-zephyr-elf/arc-zephyr-elf-gcc   -Dcheck_fstack_protector_all -nostartfiles -nostdlib -isystem /opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/arc-zephyr-elf/../../lib/arc-zephyr-elf/gcc/arc-zephyr-elf/6.2.1/include -isystem /opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/arc-zephyr-elf/../../lib/arc-zephyr-elf/gcc/arc-zephyr-elf/6.2.1/include-fixed -Wl,--unresolved-symbols=ignore-in-object-files   -fstack-protector-all -o CMakeFiles/cmTC_b2783.dir/src.c.obj   -c /projects/zephyr/tests/kernel/mem_protect/stackprot/out/CMakeFiles/CMakeTmp/src.c
Linking C executable cmTC_b2783
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_b2783.dir/link.txt --verbose=1
/opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/arc-zephyr-elf/arc-zephyr-elf-gcc  -Dcheck_fstack_protector_all -nostartfiles -nostdlib -isystem /opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/arc-zephyr-elf/../../lib/arc-zephyr-elf/gcc/arc-zephyr-elf/6.2.1/include -isystem /opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/arc-zephyr-elf/../../lib/arc-zephyr-elf/gcc/arc-zephyr-elf/6.2.1/include-fixed -Wl,--unresolved-symbols=ignore-in-object-files    CMakeFiles/cmTC_b2783.dir/src.c.obj  -o cmTC_b2783 
/opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/arc-zephyr-elf/../../libexec/arc-zephyr-elf/gcc/arc-zephyr-elf/6.2.1/real-ld: warning: cannot find entry symbol __start; defaulting to 0000000000000124
CMakeFiles/cmTC_b2783.dir/src.c.obj: In function `main':
src.c:(.text+0x10): undefined reference to `__stack_chk_guard'
src.c:(.text+0x24): undefined reference to `__stack_chk_guard'
src.c:(.text+0x2c): undefined reference to `__stack_chk_fail'
collect2: error: ld returned 1 exit status
gmake[1]: *** [CMakeFiles/cmTC_b2783.dir/build.make:98: cmTC_b2783] Error 1
gmake[1]: Leaving directory '/projects/zephyr/tests/kernel/mem_protect/stackprot/out/CMakeFiles/CMakeTmp'
gmake: *** [Makefile:126: cmTC_b2783/fast] Error 2

Source file was:
int main(void) { return 0; }
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -195,9 +195,7 @@ zephyr_cc_option(-fno-strict-overflow)
 zephyr_cc_option(-Wno-pointer-sign)
 
 if(CONFIG_STACK_CANARIES)
-  zephyr_cc_option(-fstack-protector-all)
-else()
-  zephyr_cc_option(-fno-stack-protector)
+  zephyr_cc_option_nocheck(-fstack-protector-all)
 endif()
 
 if(CONFIG_OVERRIDE_FRAME_POINTER_DEFAULT)

--- a/cmake/extensions.cmake
+++ b/cmake/extensions.cmake
@@ -103,6 +103,12 @@ function(zephyr_cc_option)
   endforeach()
 endfunction()
 
+function(zephyr_cc_option_nocheck)
+  foreach(arg ${ARGV})
+    target_compile_options(zephyr_interface INTERFACE ${arg})
+  endforeach()
+endfunction()
+
 function(zephyr_cc_option_fallback option1 option2)
     target_cc_option_fallback(zephyr_interface INTERFACE ${option1} ${option2})
 endfunction()

